liferay-frontend/liferay-portal
# LPS-136297 Enable instantiation of portlets based on registered custom elements

This builds on top of https://github.com/brianchandotcom/liferay-portal/pull/104841

----

This PR adds a new portlet to configure Custom Element (based) Portlets. These are on-the-fly portlets (don't need deployment, just configuration) that use a custom element to render their content. 

Custom Element Portlets allow sysadmins to configure the attributes to pass to the custom element (this can be used to tweak the behavior of the custom element) as well as the CSS files to use as portlet headers.

We expect to enlarge the list of configurable properties of custom element portlets in the future as we see it fits.


---

## How to test

1. Build and deploy this PR. 
2. Erase the DB
3. Start portal
4. Go to `Custom Apps` > `Remote Custom Elements` and register the [Vanilla Counter custom element](http://remote-component-test.wincent.com/packages/vanilla-counter/index.js):

![image](https://user-images.githubusercontent.com/3273630/127643802-c0333a3e-a40a-41af-965a-7f3b6ae83840.png)

5. Go to `Custom Apps` > `Custom Element Portlets` and register a new portlet based on the `vanilla-counter` custom element:

![image](https://user-images.githubusercontent.com/3273630/127643970-7015d057-0ea8-425e-8fca-f66936f20833.png)

6. Go to the home page, edit it, and add a `Vanilla Counter` portlet:

![image](https://user-images.githubusercontent.com/3273630/127644066-94691e0c-a5b6-4992-99a1-036ad6d5a2c0.png)

7. Publish the page and see it work:

![image](https://user-images.githubusercontent.com/3273630/127644119-f84af276-9a55-4aa0-ab3a-736cfedcb93a.png)

8. Make sure that the CSS file is referenced:

![image](https://user-images.githubusercontent.com/3273630/127644227-964ad6d6-e5ce-4d5f-9074-6609044b4ed7.png)

9. Make sure that the tag attributes are correctly passed to the custom element:

![image](https://user-images.githubusercontent.com/3273630/127644329-be7b5b98-aae6-469b-bbd0-a5a027ace3d3.png)
